packages/toaster/components/Toaster.tsx

```diff
@@ -15,124 +15,96 @@ interface ToasterProps {
   children?: Toast | Toast[];
 }
 
-class Toaster extends React.PureComponent<ToasterProps, { toasts: Toast[] }> {
-  public timeouts: number[] = [];
+const Toaster = ({ children }: ToasterProps) => {
+  const timeouts = React.useRef<number[]>([]);
+  const toastChildren = React.Children.toArray(children) as Toast[];
 
-  constructor(props) {
-    super(props);
+  const [toasts, setToasts] = React.useState<Toast[]>(toastChildren);
 
-    this.cloneToast = this.cloneToast.bind(this);
-    this.dismissToast = this.dismissToast.bind(this);
-    this.clearTimeouts = this.clearTimeouts.bind(this);
-    this.restartTimeouts = this.restartTimeouts.bind(this);
-    this.setTimer = this.setTimer.bind(this);
-
-    this.state = {
-      toasts: React.Children.toArray<Toast>(this.props.children).map(
-        this.cloneToast
-      )
-    };
-  }
+  const handleDismissToast = (id: ToastId = "") => {
+    const toastToDismiss = toasts.find(toast => id === toast.props.id);
+    if (toastToDismiss && toastToDismiss.props.onDismiss) {
+      toastToDismiss.props.onDismiss();
+    }
+    setToasts(toasts.filter(toast => id !== toast.props.id));
+  };
+
+  const restartTimeouts = () => {
+    toasts.forEach((toast: Toast) => {
+      const toastKey = toast.key as number;
+      if (toast.props.autodismiss) {
+        timeouts.current.push(
+          window.setTimeout(() => {
+            handleDismissToast(toast.props.id);
+          }, DELAY_TIME + MARGINAL_DELAY * toastKey)
+        );
+      }
+    });
+  };
 
-  public componentWillUnmount() {
-    this.clearTimeouts();
-  }
+  const clearTimeouts = () => {
+    timeouts.current.forEach(clearTimeout);
+  };
 
-  public componentDidMount() {
-    this.restartTimeouts();
-  }
+  React.useEffect(() => {
+    return () => {
+      clearTimeouts();
+    };
+  }, []);
 
-  public UNSAFE_componentWillReceiveProps(props: ToasterProps) {
-    const children = React.Children.toArray(props.children);
-    const childIds = children.map(toast => toast.props.id);
-    const currentIds = this.state.toasts.map(toast => toast.props.id);
+  React.useEffect(() => {
+    const childIds = toastChildren.map(toast => toast.props.id);
+    const currentIds = toasts.map(toast => toast.props.id);
 
     if (
       !currentIds.every(e => childIds.includes(e)) ||
       !currentIds.length ||
       childIds.length !== currentIds.length
```

Comment:
I feel order of this condition can be improved. 
```
!currentIds.length ||
childIds.length !== currentIds.length ||
!currentIds.every(e => childIds.includes(e))
```
This will save iterating over the loop due to short circuit if there are no child elements

Reply:
I am sure there are improvements to be made here but this effort is around refactoring from class to functional component

Reply:
the existing one is buggy however because it will create an infinite loop if both are empty arrays, so I'll modify it a bit